Biology.java

```diff
@@ -0,0 +1,397 @@
+            if (aminoAcidPreference == null) {
+                return equalPreferences;
+            }
+            return aminoAcidPreference[site];
+        }
+    }
+
+    /**
+     * Transition/transversion ratio used to determine what a given wild type nucleotide should be mutated to.
+     */
+    public enum MutationType {
+        MUTATION();
+        public final Map<Character, double[]> transitionTranversionProbability;
+
+        /**
+         * Constructor that creates a probability distribution for each nucleotide,
+         * where each possible mutation is weighted by a pre-defined transition/transversion ratio.
+         */
+        MutationType() {
+            this.transitionTranversionProbability = new HashMap<>()  {{
+                double transition = 0.5 / (Parameters.transitionTransversionRatio + 1.0);
+                double transversion = Parameters.transitionTransversionRatio / (Parameters.transitionTransversionRatio + 1.0);
+                // key: nucleotide
+                // value: int[] that gives the probability of being < to the index which corresponds to [A, C, G, T]
+
+                // Example:
+                // Parameters.transitionTransversionRatio: 5.0
+                // A: {0, 0.5/6, 5.5/6.0, 1.0}
+                // C: {0.5/6.0, 0.5/6.0, 1.0/6.0, 1.0}
+                // G: {5.0/6.0, 5.5/6.0, 5.5/6.0, 1.0}
+                // T: {0.5/6.0, 5.5/6.0, 1.0, 1.0}
+
+                put('A', new double[]{0, transition, transition + transversion, 1.0});
+                put('C', new double[]{transition, transition, transition + transition, 1.0});
+                put('G', new double[]{transversion, transversion + transition, transversion + transition, 1.0});
+                put('T', new double[]{transition, transition + transversion, 1.0, 1.0});
+            }};
+        }
+
+        /**
+         * Returns a mutant nucleotide using the transition/transversion ratio.
+         *
+         * @param originalNucleotideToMutate the original nucleotide to mutate
+         * @return the new nucleotide to change the original nucleotide to
+         */
+        public char getNucleotide(char originalNucleotideToMutate) {
+            double[] transitionTransversion = this.transitionTranversionProbability.get(originalNucleotideToMutate);
+
+            // choose a random number between 0-1
+            double randomNum = Math.random();
+            int indexAlphabet = 0;
+
+            for (int i = 0; i < 4; i++) {
+                if (randomNum < transitionTransversion[i]) {
+                    indexAlphabet = i;
+                    break;
+                }
+            }
+
+            return Biology.AlphabetType.NUCLEOTIDES.getValidCharacters().charAt(indexAlphabet);
+        }
+    }
+
+    /**
+     * DNA codon table
+     */
+    public enum CodonMap {
+        CODONS();
+        public final Map<String, String> codonMap;
+
+        CodonMap() {
+            this.codonMap = new HashMap<>() {{
```

Comment:
I trust you, Thien, but it would be nice to double check this by... perhaps printing this to CSV and sorting? Or at least checking that the AA count is right and that we have every codon represented?